### DISCUSSION TITLE
Can renovate effectively work with caching docker registry which also has local images?

### DISCUSSION THREAD

**Discussion by sergeyscherba** [renovatebot/renovate#29595] [Request Help]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub

### Please tell us more about your question or problem

Hi all

I am wondering if someone had similar experience and could advise me if renovate is capable to handle the following scenario:
We would like to use Renovate to check if there are new:
- images versions in our Helm charts
- base images versions in our Dockerfiles

**Our setup:**
- GitHub enterprise server with Renovate github action replicated to one of our repositories and then referenced in github workflow;
- Jfrog artifactory hosting private docker and helm registry. This private registry is mirroring new image/package version on-demand, when there is a try to pull this version from the clien (e.g. docker pull). New versions are not mirrorerd automatically on scheduled basis or on some other trigger.
- There is no direct access to any public registry at all

**Scenario**
- We have one of our deployments Helm template using Alpine linux 3.6 and it is set up in values.yaml;
- DockerHub already contains 3.20;
- The newest version in our private registry contains is 3.6
- We launch renovate to check if there is a new version of Alpine

Will Renovate output something like "There's no dependencies to update"? Is there any possibility to effectively use renovate in such a setup, when private registry is not really a mirror, but rather a cache?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
I'm not sure this can work if it's like you describe. If there's no direct access to public registries, and your private registries don't know about 3.20, then how could Renovate know?

**Reply by viceice** [reply]
this seems to be a limitation on artifactory. there are other discussions about mirroring.

I think artifactory only returns the local tags list instead of the upstream list, so renovate can't see them.

there's nothing renovate can do, you need to find the artifactory issue on their tracker and comment there.

maybe the issue is already resolved in newer versions of artifactory and you need to update if you self host it.

**Reply by larsskj** [reply]
We're in a similar situation: We're storing all binary artifacts, not just container images, in the JFrog Platform - in our case as a SaaS solution. Artifactory is used as a cache, and the JFrog Platform - not least X-Ray - is used to security scan binaries.

Source code is hosted in Github Enterprise, again as a SaaS solution.

I'm trying to setup Renovate using the Github app, but I've found quite a few obstacles - the problem described here being one of them.

Let me give an example: I have a bunch of Jenkins servers running in Kubernetes. For each Jenkins server I have a Git repo hosting the Kubernetes manifest files.

Renovate scans this repo and find a deployment `k8s/deploy.yaml`. This is fine.

But in this repo, it finds two Docker images to be deployed: `jenkins/jenkins` and `busybox` - the latter used for an initContainer.

Both images are loaded from our Artifactory host with an image names like `<org>.jfrog.io/<project>/jenkins/jenkins` and `<org>.jfrog.io/<project>/busybox`. Renovate correctly identifies these as Docker images with a version tag - but then complains:

```
Warning

Renovate failed to look up the following dependencies: Failed to look up docker package <org>.jfrog.io/<project>/jenkins/jenkins, Failed to look up docker package <org>.jfrog.io/<project>/busybox.

Files affected: k8s/deploy.yaml
```

That would be more or less expected. And as Artifactory is a cache, not a mirror, you cannot simply look up latest version - though pulling `latest` works, is constantly updated, and will return the latest version.

But I then tried with this configuration:

```
  "packageRules": [
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["<org>.jfrog.io/<project>/**"],
      "replacementNameTemplate": "{{{replace '<org>\\.jfrog\\.io/<project>/' 'docker.io/' packageName}}}"
    }
  ]
```

Unfortunately, it doesn't work; but if it did, problem would be solved. Or at least - I would hope so.

**Reply by viceice** [reply]
use registry aliases

**Reply by larsskj** [reply]
That certainly made a difference - but not necessarily for the better: Now I have more errors, and a bad suggestion:

`Replace <org>.jfrog.io/<project>/jenkins-exporter Docker tag with docker.io/jenkins-exporter`

That's definitely not what we want: On the contrary, we want to enforce everything to be tunneled through Artifactory.

Actually, `jenkins-exporter` is a new find that Renovate didn't detect on the first runs. And it still won't lookup the two original Docker images shown above.

**Reply by larsskj** [reply]
Ahh - it's even more complex. If I run Renovate without repo specific config _or_ with the `packageRules` described above, it just complains that it cannot find any dependencies as described above.

However, if I add registry aliases like this:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "kubernetes": {
    "fileMatch": ["^k8s/.*\\.yaml$"]
  },
  "packageRules": [
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["<org>.jfrog.io/<project>/**"],
      "replacementNameTemplate": "{{{replace '<org>\\.jfrog\\.io/<project>/' 'docker.io/' packageName}}}"
    }
  ],
  "registryAliases": {
    "docker.io": "<org>.jfrog.io/<project>",
    "docker.io/jenkins": "<org>.jfrog.io/<project>/jenkins",
    "docker.io/library": "<org>.jfrog.io/<project>"
  }
}
```

it finds the additional package `jenkins-exporter`.

However, if I _remove_ the `packageRules` section and only have the `registryAliases` section left, we're back to square one and it doesn't find the `jenkins-exporter` package.

The big difference in the Kubernetes deployment file is that the `jenkins/jenkins` and the `busybox` images are pinned with a version tag whereas the `jenkins-exporter` doesn't have a tag - it relies on `latest`.

**Reply by larsskj** [reply]
Reverting the order of the registry aliases did the trick: Now it actually works!

It correctly suggest updates to Jenkins and Busybox from Docker Hub, but it also preserves the original image name in the Kubernetes deployment - meaning the updates will be pulled from Artifactory, not Docker hub. Cudos!

This works:

```
  "registryAliases": {
    "<org>.jfrog.io/<project>": "docker.io",
    "<org>.jfrog.io/<project>/jenkins": "docker.io/jenkins",
    "<org>.jfrog.io/<project>/library": "<org>.jfrog.io/<project>/library"
  }
```

**Reply by HonkingGoose** [reply]
The [`registryAliases` docs](https://docs.renovatebot.com/configuration-options/#registryaliases) do _not_ mention that the order matters. I guess Renovate checks the registry aliases from top to bottom?

So, given this input:

```
  "registryAliases": {
    "<org>.jfrog.io/<project>": "docker.io",
    "<org>.jfrog.io/<project>/jenkins": "docker.io/jenkins",
    "<org>.jfrog.io/<project>/library": "<org>.jfrog.io/<project>/library"
  }
```

Renovate will:

1. Check `docker.io` first?
1. Then `docker.io/jenkins`?
1. Lastly `<org>.jfrog.io/<project>/library`?

**Reply by larsskj** [reply]
Yeah - sorry about the confusion: I had a very busy day with a lot of disturbances at work yesterday.

The line `"<org>.jfrog.io/<project>/library": "<org>.jfrog.io/<project>/library"` was a mistake - it was meant to be `"<org>.jfrog.io/<project>/library": "docker.io/library"`.

And further experiments proved that it wasn't required: The first line is enough for our setup to work.

I added all three lines as an attempt to cover all eventualities until I had a working solution. I don't find the documentation very informative in this regard - but then again, I'm very new to Renovate.

**Reply by HonkingGoose** [reply]
> I added all three lines as an attempt to cover all eventualities until I had a working solution.

Hmm... Maintainers: does the order of items in `registryAliases` _matter_ make a difference? Or is it enough to have all the wanted aliases in the list, even if the order is random?

> I don't find the documentation very informative in this regard - but then again, I'm very new to Renovate.

I hope to improve the `registryAliases` part of the docs, once I know what to put there. :wink:

**Reply by viceice** [reply]
they are all applied from top to bottom

**Reply by larsskj** [reply]
Thank you for the clarification.

Is that documented somewhere?

**Reply by HonkingGoose** [reply]
> Is that documented somewhere?

I'm putting it in the docs, with this PR:

- #30691

**Reply by juv** [reply]
@viceice 
> this seems to be a limitation on artifactory. there are other discussions about mirroring.
> 
> I think artifactory only returns the local tags list instead of the upstream list, so renovate can't see them.
> 
> there's nothing renovate can do, you need to find the artifactory issue on their tracker and comment there.
> 
> maybe the issue is already resolved in newer versions of artifactory and you need to update if you self host it.

From a first glance, it looks like this behaviour has changed in the meantime? 
I have found this ticket which is closed as `done`: https://jfrog.atlassian.net/browse/RTFACT-29270 

and also this [documentation](https://jfrog.com/help/r/artifactory-how-to-resolve-packages-when-it-is-removed-from-the-upstream-remote-repository/understanding-artifactory-s-metadata-handling), which says: 
> Artifactory efficiently organizes and tracks packages through metadata. The expiration of the "Metadata Retrieval Cache Period" prompts Artifactory to fetch updated metadata files from the upstream repository. If new versions are introduced, querying for a package without specifying the version retrieves the latest version from the upstream.

Do you think it is worth trying it out again? I am most interested in the use case where I have a `Dockerfile` that is pointing to my private Artifactory URL as the `base image` and would like Renovatebot to update this base image version in case there is something new. 

Example: 
Dockerfile contains
`FROM my.private.jfrog.com/remote-docker-io/maven:3.9.9-eclipse-temurin-21-alpine AS builder` 
and I would like to see Renovate pull requests to this Dockerfile for new versions of Maven, e.g. `3.9.10-eclipse-temurin-21-alpine`

**Reply by rarkins** [reply]
It's worth testing again - please go ahead. If JFrog has fixed this as you hope, you should see the new PR appearing. If you don't, they probably haven't fixed it.

**Comment by larsskj** [comment]
More fuel for this discussion:

I've now spent much more time trying to make Renovate work with our setup - but I cannot figure it out.

If `registryAliases` really is what the name implies, I guess it's working as designed - but in that case I guess we need to request an enhancement.

Let me give an example: We have a lot of Jenkins servers running in Kubernetes. I've created a repo on Github for each server, and let Renovate process these repos. Each repo basically holds the K8s manifest files and a Jenkinsfile that will re-deploy the Jenkins master when the deployment manifest is updated.

The Jenkins pods running our servers pulls three container images: A `busybox` image running as an `initContainer` do initialize the workspace, and then the `jenkins` master container and a sidecar container running a special `akawork-jenkins-exporter` that will add additional datasets for Prometheus to scrape.

`busybox` and `jenkins` are unmodified downloads from Docker Hub, while the `akawork-jenkins-exporter` is a fork of a public Github repo that has been stale for five years. We've forked it, updated it, and rebuilt it - and host the result in our private Artifactory repository.

This is where the fun starts: For security reasons, _all_ container images are pulled from the private Artifactory instance which not only caches the images, but also has the full X-Ray security package for monitoring all container images being deployed within the organization.

Renovate looks for container images in the repo's `k8s/deploy.yaml` file and correctly identifies the three container images. I have created a `renovate.json` with the following content:

```
  "registryAliases": {
    "<org>.jfrog.io/<registry>": "docker.io"
  }
```

and that works in the respect that the full name of the three images are modified from, say, `<org>.jfrog.io/<registry>/jenkins` to `docker.io/jenkins`. Renovate looks up the relevant data for `busybox` and `jenkins` - and complains that `akawork-jenkins-exporter` doesn't exist on Docker Hub. Which is perfectly true.

I've then tried

```
  "registryAliases": {
    "<org>.jfrog.io/<registry>/akawork-jenkins-exporter": "<org>.jfrog.io/<registry>/akawork-jenkins-exporter",
    "<org>.jfrog.io/<registry>": "docker.io"
  }
```

but all three images still points to Docker Hub. I had hoped that it would process the first hit and then stop processing - but that's not the case. So I tried

```
  "registryAliases": {
    "<org>.jfrog.io/<registry>": "docker.io",
    "<org>.jfrog.io/<registry>/akawork-jenkins-exporter": "<org>.jfrog.io/<registry>/akawork-jenkins-exporter"
  }
```

but again to no avail. Then I thought - OK, we're processing top to bottom, so maybe

```
  "registryAliases": {
    "<org>.jfrog.io/<registry>": "docker.io",
    "docker.io/akawork-jenkins-exporter": "<org>.jfrog.io/<registry>/akawork-jenkins-exporter"
  }
```

but neither does this work.

This can become much more complex as we start referring images "living" in other registries than Docker Hub - which is very much the case nowadays.

Is there a way to configure this in Renovate?

**Reply by rarkins** [reply]
In artifactory terminology, you have a single repository (Registry) which mixes both "local" (your own images) and "remote" (docker hub images)? If you weren't mixing them then you could apply registryAliases to your remote only, and leave local as-is.

It might still be possible for you to use packageRules to selectively apply `registryUrls` to your internal packages. `registryUrls` only work when the image host is Docker Hub, although I'm not sure if that also works when it's like `docker.io/something` instead of just `something`

**Reply by larsskj** [reply]
> In artifactory terminology, you have a single repository (Registry) which mixes both "local" (your own images) and "remote" (docker hub images)? If you weren't mixing them then you could apply registryAliases to your remote only, and leave local as-is.

It's a bit more complicated:

You have local registries - in plural. You can have different registries for different projects, and you will often have different registries for development and production repositories. You can then promote from, say, `dev` to `prod` with different retention policies, usage policies, etc.

And you have remote registries: These will usually point to various public registries that are then cached (not mirrored) in Artifactory.

Finally, you have virtual registries: These are normally created by combining local and remote registries into a single, unified registry. Users then need to point to this virtual registry only, hiding the complexities of the setup. When a public project moves from one registry to another - like projects moving away from Docker Hub - this is handled internally by Artifactory admins, whereas end users pointing to the virtual registry do not need to be aware of the change.

You can point to individual physical registries instead of the virtuals, but that's against recommended practice.
 
> It might still be possible for you to use packageRules to selectively apply `registryUrls` to your internal packages. `registryUrls` only work when the image host is Docker Hub, although I'm not sure if that also works when it's like `docker.io/something` instead of just `something`

It would be amazing if a solution for Renovate could be found, as I believe still more people will look into solutions like the Artifactory setup described above.

And the basic Artifactory setup described above is not only valid for container images, the same setup basically applies to all package types supported by the JFrog platform. And that's a lot.
